### ISSUE TITLE
Add exponential moving average (EMA) of model weights

### ISSUE THREAD

**Issue by swfsql** [tracel-ai/burn#5402]
_(Prepared by Claude)._

## What

A first-class way to keep an exponentially-moving average of a module's float
parameters during training, and to evaluate / checkpoint from it:

```
shadow ← decay · shadow + (1 − decay) · param      (after each optimizer step)
```

This can be hand-rolled per project on top of `ModuleMapper`, though a few
details are easy to miss: the shadow wants to be detached rather than
autodiff-tracked, to follow the model's device/dtype, to warm the decay up so it
isn't anchored to the initialization, and to participate in `Record` so a
resumed run keeps it.

If this seems worth having, `burn-optim` next to `lr_scheduler` and
`grad_clipping` might be a reasonable home, perhaps with `LearnerBuilder`
integration in `burn-train`.

Useful implementation references: PyTorch
(`torch.optim.swa_utils.AveragedModel` + `get_ema_multi_avg_fn`), Optax
(`optax.ema`), Keras 3 (`use_ema` / `ema_momentum` on `Optimizer`), timm
(`ModelEmaV3`), HF Diffusers (`EMAModel`).

## Research references

- **Polyak & Juditsky (1992)**, *Acceleration of Stochastic Approximation by
  Averaging*, SIAM J. Control Optim. 30(4) — the classical result that averaging
  the iterates improves the convergence rate of stochastic approximation.
  Uniform averaging rather than exponential, but the origin of the idea.
- **Izmailov et al. (2018)**, *Averaging Weights Leads to Wider Optima and Better
  Generalization* (SWA), UAI 2018 — averaged weights find much flatter solutions
  than SGD and generalize better than the final iterate. Also uniform averaging;
  EMA is the online, constant-memory variant that the works below use.
- **Tarvainen & Valpola (2017)**, *Mean teachers are better role models*,
  NeurIPS 2017 — an EMA of the weights as a stable target, now standard in
  semi-supervised and self-supervised training.
- **Karras et al. (2024)**, *Analyzing and Improving the Training Dynamics of
  Diffusion Models* (EDM2), CVPR 2024 — treats the EMA length as a first-class
  hyperparameter with a measurable optimum, and contributes a method for setting
  the EMA parameters *post-hoc*, after training has finished.
- **Jolicoeur-Martineau (2025)**, *Less is More: Recursive Reasoning with Tiny
  Networks* (TRM) — ablation on Sudoku-Extreme: 79.9% → 87.4% test accuracy from
  EMA (decay 0.999) alone; without it the run collapses sharply and then
  diverges. One of the three largest effects in the paper's ablation table.
- **HRM-Text (2026)** — uses EMA (decay 0.9999) *instead of* learning-rate decay
  for a 1B-parameter run; the reported evaluations and the publicly released
  weights are the EMA checkpoint.

The last two are recent small-budget training runs where EMA mattered for
stability rather than as a final cosmetic gain, which is part of what prompted
this.

---

- _Edit: remade the prose, which was weirdly unnecessarily aggressive._
